### PR TITLE
⚠️ Allow specifying a different credentials per VSphereMachine

### LINKED ISSUE #1720: [WIP] ⚠️ Allow specifying a different credentials per VSphereMachine
/kind feature

**Describe the solution you'd like**
It would be great if CAPV would support multiple vCenter in one cluster.

Use Cases:

- Migration between vCenter's
- Blast reduce reduction

Any idea or other use cases?

### PR DESCRIPTION
**Pull request by farodin91**
**What this PR does / why we need it**:
Allow multiple vcenter in a single cluster

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1720

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

I will try to test it.

**Release note**:

```release-note
Allow custom identity in vmmachine
```

### CODE REVIEW

**Comment by farodin91** [comment]
We are using this commit to migrate cluster from one vcenter to a different and it works just fine.

**Comment by farodin91** [comment]
@MaxRink would you like to look into this PR?

**Comment by farodin91** [comment]
@srm09 would you like to look into this PR?

**Comment by randomvariable** [comment]
We discussed this in the call of 2023/07/23:

* This will need to be marked as a breaking change, as there is a security implication of allowing users to create MachineDeployments targeting differing vcenters.
* We need an expansion of docs about how this works in a cluster and how to set up CPI and CSI in a cluster to  make it work

cc @puneetkatyal

**Review by randomvariable — CHANGES_REQUESTED** [review]
As per https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/pull/1743#issuecomment-1644302112

**Comment by farodin91** [comment]
@randomvariable  I started to write a documentation. 

I added a link to external documentation for csi and cpi, do you think this is enough?

How should i mark this a breaking change?

**Comment by chrischdi** [comment]
/retitle ⚠️ Allow specifying a different credentials per VSphereMachine

(warning because of the security impact)

**Comment by farodin91** [comment]
@randomvariable Should i add this to commit message?

**Comment by guilhermevillote** [comment]
Any updates on this PR? This is something very useful when your infrastructure has multiple vcenters and you constantly needs to migrate machines from one to another..

Thanks

**Comment by guilhermevillote** [comment]
Hi, any updates on this PR?

File: pkg/identity/identity_test.go [hunk 1]
```diff
@@ -236,19 +236,19 @@ var _ = Describe("validateInputs", func() {
 
 	Context("If the client is missing", func() {
 		It("should error if client is missing", func() {
-			Expect(validateInputs(nil, cluster)).NotTo(Succeed())
```
Comment (chrischdi) [inline pkg/identity/identity_test.go:239]: would need test cases for the new behavior

File: apis/v1alpha3/zz_generated.conversion.go [hunk 2]
```diff
@@ -1687,5 +1687,6 @@ func autoConvert_v1beta1_VirtualMachineCloneSpec_To_v1alpha3_VirtualMachineClone
 	// WARNING: in.PciDevices requires manual conversion: does not exist in peer-type
 	// WARNING: in.OS requires manual conversion: does not exist in peer-type
 	// WARNING: in.HardwareVersion requires manual conversion: does not exist in peer-type
+	// WARNING: in.IdentityRef requires manual conversion: does not exist in peer-type
```
Comment (chrischdi) [inline apis/v1alpha3/zz_generated.conversion.go:1690]: v1alpha3 is gone on main (similar v1alpha4?)

File: docs/multi-vcenter.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
+# Multi VCenter support
+
+Cluster API Provider vSphere (CAPV) supports multiple VCenter for a single. Therefore CAPV is allowing to define the used identity for each machine. CAPV will check on every Machine first, if there is a local identity otherwise it fallback on the default selection method. 
+
+In order to run a CAPV cluster in multiple VCenter, you have to configure CPI & CSI to support multi VCenter, see [guide](https://docs.vmware.com/en/VMware-vSphere-Container-Storage-Plug-in/3.0/vmware-vsphere-csp-getting-started/GUID-8B3B9004-DE37-4E6B-9AA1-234CDA1BD7F9.html). Trivia, `VSphereCluster` can be only in single VCenter. This will just used as a fallback, if you haven't configured a different identity for a `VSphereMachine``.
```
Comment (chrischdi) [inline docs/multi-vcenter.md:5]: This only links csi, are there also instructions for CPI?

File: pkg/identity/identity.go [hunk 4]
```diff
@@ -108,15 +107,21 @@ func GetCredentials(ctx context.Context, c client.Client, cluster *infrav1.VSphe
 	return credentials, nil
 }
 
-func validateInputs(c client.Client, cluster *infrav1.VSphereCluster) error {
+func GetCredentials(ctx context.Context, c client.Client, cluster *infrav1.VSphereCluster, controllerNamespace string) (*Credentials, error) {
```
Comment (chrischdi) [inline pkg/identity/identity.go:110]: unit tests would be helpful

File: docs/multi-vcenter.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# Multi VCenter support
+
+Cluster API Provider vSphere (CAPV) supports multiple VCenter for a single. Therefore CAPV is allowing to define the used identity for each machine. CAPV will check on every Machine first, if there is a local identity otherwise it fallback on the default selection method.
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply docs/multi-vcenter.md:3]: Also what does the first sentence mean?

> Cluster API Provider vSphere (CAPV) supports multiple VCenter for a single.

File: docs/multi-vcenter.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# Multi VCenter support
```
Comment (chrischdi) [inline docs/multi-vcenter.md:1]: Is most of this better suited to Also should ahve some explanation how it works in https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/docs/identity_management.md ?

File: controllers/vspherevm_controller.go [hunk 7]
```diff
@@ -554,29 +554,25 @@ func (r vmReconciler) retrieveVcenterSession(ctx goctx.Context, vsphereVM *infra
...
-		Name:      cluster.Spec.InfrastructureRef.Name,
-	}
-	vsphereCluster := &infrav1.VSphereCluster{}
-	err = r.Client.Get(r, key, vsphereCluster)
-	if err != nil {
-		r.Logger.Info("VSphereCluster couldn't be retrieved")
-		return session.GetOrCreate(r.Context,
-			params)
+	if vsphereVM.Spec.IdentityRef != nil {
+		creds, err := identity.GetCredentialsWithExternalIdentity(ctx, r.Client, vsphereCluster, vsphereVM.Spec.IdentityRef, r.Namespace)
```
Comment (chrischdi) [inline controllers/vspherevm_controller.go:569]: Wouldn't this get overwritten below in l 578 when there is a definition on the cluster level already?

**Review by chrischdi — COMMENTED** [review]
some comments, definetly needs a rebase too.

**Comment by farodin91** [comment]
@chrischdi @guilhermevillote If someone like to take over i'm fine. I don't have time at moment to fix it.

**Comment by chrischdi** [comment]
> @chrischdi @guilhermevillote If someone like to take over i'm fine. I don't have time at moment to fix it.

Thanks @farodin91 !

/help

**Pull request closed without merging**